### ISSUE TITLE
spack mirror command not working

### ISSUE THREAD

**Issue by jgalarowicz** [LLNL/spack#34532]
### Steps to reproduce

I was trying to build survey on snow at LANL using a mirror with the latest 12-14-2022 spack sources, but I get this message.   I was able to use the same mirror with an older 11-21-2022 version of the spack sources.   So, thought I would report this.
```
sn-fe1:/<4>survey/spack_12142022/spack [841] $ spack install --no-cache --keep-stage [survey@1.0.6.release-openmpi-2.1.2](mailto:survey@1.0.6.release-openmpi-2.1.2) %gcc@8.3.0 +mpi ^openmpi@2.1.2
==> Error: unknown url type: 'usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022/build_cache/index.json'
```

### Error message

```
sn-fe1:/<4>survey/spack_12142022/spack [841] $ spack install --no-cache --keep-stage [survey@1.0.6.release-openmpi-2.1.2](mailto:survey@1.0.6.release-openmpi-2.1.2) %gcc@8.3.0 +mpi ^openmpi@2.1.2
==> Error: unknown url type: 'usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022/build_cache/index.json'
```

### Information on your system

```
$ spack debug report
* **Spack:** 0.20.0.dev0 (2387c116add454d6b7cf5400e368fb95bd3de415)
* **Python:** 3.6.8
* **Platform:** linux-rhel7-broadwell
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by haampie** [comment]
Can you show what `spack config get mirrors` looks like?

It's likely related to e055dc0e64693782e5d98ffa90780ad57c855aa2. The issue being solved there is Windows support; in the past we've been using URLs and paths interchangeably, but that results in various issues, such as c:/users/x/y being interpreted as a URL like http://users/x/y where it should have been a windows path etc.

We can try to promote the path better to a URL.

**Comment by jgalarowicz** [comment]
@haampie 
Here is the information:
```
sn-fe1:/<4>survey/spack_12142022/spack [948] $ spack config get mirrors
mirrors:
  survey-mirror-12-14-2022:
    fetch:
      url: /usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022
      access_pair:
      - null
      - null
      access_token: null
      profile: null
      endpoint_url: null
    push:
      url: /usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022
      access_pair:
      - null
      - null
      access_token: null
      profile: null
      endpoint_url: null
  spack-public: https://mirror.spack.io
```

**Comment by jgalarowicz** [comment]
Here is the debug output:

```
sn-fe1:/<4>survey/spack_12142022/spack [956] $ spack -d install --no-cache --keep-stage survey@1.0.6.release-openmpi-4.1.1 %gcc@8.3.0 +mpi ^openmpi@4.1.1
==> [2022-12-14-14:37:23.857728] Imported install from built-in commands
==> [2022-12-14-14:37:23.860397] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/config.yaml
==> [2022-12-14-14:37:23.886493] Reading config from file /users/jegsgi/.spack/config.yaml
==> [2022-12-14-14:37:23.890071] Imported install from built-in commands
==> [2022-12-14-14:37:23.901176] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-12-14-14:37:23.916325] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/repos.yaml
==> [2022-12-14-14:37:23.923129] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/config.yaml
==> [2022-12-14-14:37:23.948373] Reading config from file /users/jegsgi/.spack/config.yaml
==> [2022-12-14-14:37:23.951329] DATABASE LOCK TIMEOUT: 3s
==> [2022-12-14-14:37:23.951414] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-12-14-14:37:23.953313] DATABASE LOCK TIMEOUT: 3s
==> [2022-12-14-14:37:23.953401] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-12-14-14:37:23.953527] [BOOTSTRAP CONFIG SCOPE] name=_builtin
==> [2022-12-14-14:37:23.954035] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-12-14-14:37:23.963844] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults
==> [2022-12-14-14:37:23.963933] [BOOTSTRAP CONFIG SCOPE] name=defaults/linux, path=/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/linux
==> [2022-12-14-14:37:23.964009] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/users/jegsgi/.spack/bootstrap/config
==> [2022-12-14-14:37:23.964057] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/linux, path=/users/jegsgi/.spack/bootstrap/config/linux
==> [2022-12-14-14:37:23.965010] Reading config from file /users/jegsgi/.spack/bootstrap/config/linux/compilers.yaml
==> [2022-12-14-14:37:24.012131] [BOOTSTRAP ROOT SPEC] patchelf@0.13.1: %gcc target=x86_64
==> [2022-12-14-14:37:24.012709] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/bootstrap.yaml
==> [2022-12-14-14:37:24.029542] [BOOTSTRAP EXECUTABLES patchelf] Try installed specs with query 'patchelf@0.13.1: %gcc target=x86_64'
==> [2022-12-14-14:37:25.985613] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/config.yaml
==> [2022-12-14-14:37:26.092365] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/modules.yaml
==> [2022-12-14-14:37:26.110499] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/linux/modules.yaml
==> [2022-12-14-14:37:26.265344] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/packages.yaml
==> [2022-12-14-14:37:26.307733] Reading config from file /users/jegsgi/.spack/bootstrap/config/packages.yaml
==> [2022-12-14-14:37:26.320053] [BOOTSTRAP ROOT SPEC] gnupg@2.3: %gcc target=x86_64
==> [2022-12-14-14:37:26.325140] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %gcc target=x86_64
==> [2022-12-14-14:37:26.325231] [BOOTSTRAP MODULE clingo] Try importing from Python
==> [2022-12-14-14:37:26.390941] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %gcc target=x86_64 ^python@3.6/vcipwnf57slgoo7busvvkzjkk7vydeb5" provides the "clingo" Python module
==> [2022-12-14-14:37:26.391886] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/concretizer.yaml
==> [2022-12-14-14:37:26.398387] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/repos.yaml
==> [2022-12-14-14:37:26.474187] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/packages.yaml
==> [2022-12-14-14:37:26.516195] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/packages.yaml
==> [2022-12-14-14:37:26.538908] Reading config from file /users/jegsgi/.spack/linux/packages.yaml
==> [2022-12-14-14:37:26.586420] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/config.yaml
==> [2022-12-14-14:37:26.630650] Reading config from file /users/jegsgi/.spack/config.yaml
==> [2022-12-14-14:37:26.637996] Reading config from file /turquoise/usr/projects/hpctools/survey/spack_12142022/spack/etc/spack/defaults/mirrors.yaml
==> [2022-12-14-14:37:26.640174] Reading config from file /users/jegsgi/.spack/mirrors.yaml
==> [2022-12-14-14:37:26.647266] Checking existence of usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022/build_cache/index.json
Traceback (most recent call last):
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack_installable/main.py", line 36, in main
    sys.exit(spack.main.main(argv))
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/main.py", line 1000, in main
    return _main(argv)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/main.py", line 955, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/main.py", line 983, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/main.py", line 632, in _invoke_command
    return_val = command(parser, args)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/cmd/install.py", line 531, in install
    abstract_specs, concrete_specs = specs_from_cli(args, install_kwargs, reporter)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/cmd/install.py", line 448, in specs_from_cli
    args.spec, concretize=True, tests=install_kwargs["tests"]
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/cmd/__init__.py", line 224, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/spec.py", line 2950, in concretize
    self._new_concretize(tests)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/spec.py", line 2925, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/solver/asp.py", line 2515, in solve
    reusable_specs.extend(self._reusable_specs())
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/solver/asp.py", line 2484, in _reusable_specs
    index = spack.binary_distribution.update_cache_and_get_specs()
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/binary_distribution.py", line 2122, in update_cache_and_get_specs
    binary_index.update()
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/binary_distribution.py", line 402, in update
    needs_regen = self._fetch_and_cache_index(mirror_url)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/binary_distribution.py", line 448, in _fetch_and_cache_index
    if not web_util.url_exists(index_fetch_url):
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/util/web.py", line 409, in url_exists
    read_from_url(url)
  File "/turquoise/usr/projects/hpctools/survey/spack_12142022/spack/lib/spack/spack/util/web.py", line 105, in read_from_url
    req = Request(url, headers={"User-Agent": SPACK_USER_AGENT})
  File "/var/lib/perceus/vnfs/fe/rootfs/usr/lib64/python3.6/urllib/request.py", line 329, in __init__
    self.full_url = url
  File "/var/lib/perceus/vnfs/fe/rootfs/usr/lib64/python3.6/urllib/request.py", line 355, in full_url
    self._parse()
  File "/var/lib/perceus/vnfs/fe/rootfs/usr/lib64/python3.6/urllib/request.py", line 384, in _parse
    raise ValueError("unknown url type: %r" % self.full_url)
ValueError: unknown url type: 'usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022/build_cache/index.json'
sn-fe1:/<4>survey/spack_12142022/spack [957] $ 
```

**Comment by haampie** [comment]
Thanks, in this case there's indeed a path where a url is requested. You can turn it into a url:

```
$ spack config get mirrors
mirrors:
  survey-mirror-12-14-2022: file:///usr/projects/hpctools/survey/spack_12142022/survey-mirror-12-14-2022
  spack-public: https://mirror.spack.io
```

should fix it. I'll submit a PR that turns this in a warning & updates the config automatically.

(Actually, it seems that `spack mirror add` is also asking for a url, but doesn't validate input.)

**Comment by jgalarowicz** [comment]
Thank you for looking at this and for the solution!

**Comment by lpoirel** [comment]
Hello, I seem to get a very similar issue when trying to bootstrap following https://spack.readthedocs.io/en/latest/bootstrapping.html#creating-a-mirror-for-air-gapped-systems

Should I file it as a separate issue ?

The fix of replacing "/path/to/my/mirror" with "file:///path/to/my/mirror" does not work: the "file" prefix is taken as a relative path as shown below:
```
$ cat ~/.spack/bootstrap.yaml
bootstrap:
  sources:
  - name: local-sources
    metadata: file:///home/lpoirel/mirror_htop/metadata/sources
  - name: local-binaries
    metadata: file:///home/lpoirel/mirror_htop/metadata/binaries
  trusted:
    local-binaries: true
    local-sources: true
$ spack bootstrap now
Traceback (most recent call last):
  File "/home/lpoirel/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/home/lpoirel/spack/lib/spack/spack_installable/main.py", line 36, in main
    sys.exit(spack.main.main(argv))
  File "/home/lpoirel/spack/lib/spack/spack/main.py", line 1000, in main
    return _main(argv)
  File "/home/lpoirel/spack/lib/spack/spack/main.py", line 955, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/lpoirel/spack/lib/spack/spack/main.py", line 983, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/lpoirel/spack/lib/spack/spack/main.py", line 632, in _invoke_command
    return_val = command(parser, args)
  File "/home/lpoirel/spack/lib/spack/spack/cmd/bootstrap.py", line 462, in bootstrap
    callbacks[args.subcommand](args)
  File "/home/lpoirel/spack/lib/spack/spack/cmd/bootstrap.py", line 444, in _now
    spack.bootstrap.ensure_core_dependencies()
  File "/home/lpoirel/spack/lib/spack/spack/bootstrap/core.py", line 547, in ensure_core_dependencies
    ensure_patchelf_in_path_or_raise()
  File "/home/lpoirel/spack/lib/spack/spack/bootstrap/core.py", line 534, in ensure_patchelf_in_path_or_raise
    executables=["patchelf"], abstract_spec=patchelf_root_spec(), cmd_check=verify_patchelf
  File "/home/lpoirel/spack/lib/spack/spack/bootstrap/core.py", line 427, in ensure_executables_in_path_or_raise
    for current_config in bootstrapping_sources():
  File "/home/lpoirel/spack/lib/spack/spack/bootstrap/core.py", line 572, in bootstrapping_sources
    with open(metadata_yaml, encoding="utf-8") as stream:
FileNotFoundError: [Errno 2] No such file or directory: '/home/lpoirel/.spack/file:/home/lpoirel/mirror_htop/metadata/sources/metadata.yaml'
```

**Comment by haampie** [comment]
Thanks @lpoirel, should also be fixed in #34552

**Comment by white238** [comment]
I hit this today when trying to update Serac. Thank you for already having a fix in the works!

**Comment by haampie** [comment]
The PR went through a second iteration, it's now #34891, and "old" config where mirrors are stored as paths will work again... sorry for the inconvenience and delay in merging.

**Comment by alalazo** [comment]
Fixed by #34891
